02_bubble/brnbke/main.go

```diff
@@ -0,0 +1,91 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+)
+
+var out io.Writer = os.Stdout
+
+func bubble(arr []int) []int {
+	changed := true
+	r := make([]int, len(arr))
+	copy(r, arr)
+
+	for {
+		if !changed {
+			break
+		}
+		changed = false
+		for i := 0; i < len(r)-1; i++ {
+			if r[i] > r[i+1] {
+				r[i], r[i+1] = r[i+1], r[i]
+				changed = true
+			}
+		}
+	}
+
+	return r
+}
+
+func insertion(arr []int) []int {
+	result := make([]int, len(arr))
+	copy(result, arr)
+
+	for i := 1; i < len(result); i++ {
+		tmp := result[i]
+		for j := i - 1; j > -1; j-- {
+			if result[j] > tmp {
+				result[j+1], result[j] = result[j], tmp
+			} else {
+				break
+			}
+		}
+	}
+
+	return result
+}
+
+func mergesort(arr []int) []int {
+	if len(arr) <= 1 {
+		return arr
+	}
+
+	// work out the mid point of array
+	i := len(arr) / 2
+
+	// split input arr into two halves and recursively call merge sort to sort each half
+	l := mergesort(arr[:i])
+	r := mergesort(arr[i:])
+
+	// merge two sorted halves back together.
+	return merge(l, r)
+}
+
+func merge(l []int, r []int) []int {
+
+	result := []int{}
```

Comment:
Given that you know the size of the result (`len(l)+len(r)`), you could create a pre-allocated array here. Then you could guarantee that you never need to regrow the array through a call to `append`.

Reply:
Done